### PR TITLE
Kademlia Dialer

### PR DESCRIPTION
**Pull request by egonelbre**

### CODE REVIEW

File: pkg/kademlia/pool.go [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+	limit  sync2.Semaphore
+	mu     sync.Mutex
+	closed bool
+	recent []*Conn
+}
+
+// Conn represents a cached kademlia cache.
+type Conn struct {
+	pool     *Pool
+	refcount int32 // only modify when holding pool.mu
```
Comment (jtolds) [inline pkg/kademlia/pool.go]: oh whoa uh, okay

Comment (jtolds) [reply pkg/kademlia/pool.go]: i feel like the right thing here is to have conn not have the refcount, but have Pool have a mapping of Conns to refcounts internally

File: pkg/kademlia/pool.go [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+
+	for i, conn := range pool.recent {
+		// TODO: verify also other properties
+		if conn.addr == target.GetAddress().Address {
+			conn.acquireLocked()
+
+			// this is a simplest way to implement recently used cache
+			// avoiding random connections pushing important connections off
+			k := i / 2
+			pool.recent[k], pool.recent[i] = pool.recent[i], pool.recent[k]
```
Comment (jtolds) [inline pkg/kademlia/pool.go]: neat

File: pkg/kademlia/dialer.go [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+	}
+	defer dialer.limit.Unlock()
+
+	conn, err := dialer.dial(ctx, ask)
+	if err != nil {
+		return nil, err
+	}
+
+	resp, err := conn.client.Query(ctx, &pb.QueryRequest{
+		Limit:    20,
```
Comment (jtolds) [inline pkg/kademlia/dialer.go]: not part of this PR but this should absolutely be passed in and not hardcoded. this should be the kademlia k value, routing table depth, etc

File: pkg/kademlia/dialer.go [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+		Sender:   &self,
+		Target:   &find,
+		Pingback: true,
```
Comment (jtolds) [inline pkg/kademlia/dialer.go]: not part of this PR but pingback should not always be true. in fact, it should only be true during routine kademlia specific bucket refreshes

File: pkg/kademlia/dialer.go (hunk 3 above)
Comment (egonelbre) [reply pkg/kademlia/dialer.go]: yeah, noticed this as well, I'll add a TODO

File: pkg/kademlia/dialer.go (hunk 4 above)
Comment (egonelbre) [reply pkg/kademlia/dialer.go]: I'll add a TODO

File: pkg/kademlia/dialer.go [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+	})
+	if err != nil {
+		return nil, errs.Combine(err, conn.disconnect())
+	}
+
+	return resp.Response, errs.Combine(err, conn.disconnect())
```
Comment (thepaul) [inline pkg/kademlia/dialer.go]: err is guaranteed to be nil here. still call errs.Combine?
